sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts

```diff
@@ -673,17 +707,198 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
+      .addSelect("application.data->>'workflowName'", "workflowName")
+      .addSelect("assessment.assessmentWorkflowId", "assessmentWorkflowId")
+      .innerJoin("application.currentAssessment", "assessment")
+      .innerJoin(
+        EducationProgramOffering,
+        "offering",
+        "offering.precedingOffering.id = assessment.offering.id",
+      )
+      .where("application.applicationStatus != :cancelledStatus", {
+        cancelledStatus: ApplicationStatus.cancelled,
+      })
+      .andWhere("offering.id = :offeringId", { offeringId })
+      .getRawAndEntities();
+    return mapFromRawAndEntities<ApplicationAssessmentSummary>(
+      applications,
+      "workflowName",
+      "assessmentWorkflowId",
+      "hasAssessmentData",
+    );
+  }
+
+  /**
+   * Approve or Decline an offering change
+   * requested by institution.
+   * @param offeringId offering that is requested for change.
+   * @param userId User who approves or declines the offering.
+   * @param assessmentNotes Notes added during the process.
+   * @param offeringStatus Approval or Decline status.
+   */
+  async assessOfferingChangeRequest(
+    offeringId: number,
+    userId: number,
+    assessmentNotes: string,
+    offeringStatus: OfferingStatus,
+  ): Promise<void> {
+    const requestedOffering = await this.getOfferingSummaryToRequestChange(
+      offeringId,
+      OfferingStatus.AwaitingApproval,
+    );
+
+    if (!requestedOffering) {
+      throw new CustomNamedError(
+        `Either offering not found or the offering not in appropriate status to be approved or declined for change.`,
+        OFFERING_NOT_VALID,
+      );
+    }
+
+    if (!requestedOffering.precedingOffering) {
+      throw new CustomNamedError(
+        `The offering requested for change does not have a preceding offering.`,
+        OFFERING_NOT_VALID,
+      );
+    }
+    const precedingOffering = requestedOffering.precedingOffering;
+    const auditUser = { id: userId } as User;
+    const currentDate = new Date();
+    //Set audit fields.
+    requestedOffering.modifier = auditUser;
+    requestedOffering.updatedAt = currentDate;
+    requestedOffering.assessedBy = auditUser;
+    requestedOffering.assessedDate = currentDate;
+
+    precedingOffering.modifier = auditUser;
+    precedingOffering.updatedAt = currentDate;
+    precedingOffering.assessedBy = auditUser;
+    precedingOffering.assessedDate = currentDate;
+    let applications: ApplicationAssessmentSummary[] = [];
+
+    //Populate the status accordingly and get impacted applications
+    // when the offering change is approved.
+    if (offeringStatus === OfferingStatus.Approved) {
+      requestedOffering.offeringStatus = OfferingStatus.Approved;
+      precedingOffering.offeringStatus = OfferingStatus.ChangeOverwritten;
+      applications = await this.getApplicationsToSubmitReassessment(offeringId);
+
+      for (const application of applications) {
```

Comment:
I'm thinking it will be valuable to check if there are applications (applications.length > 0) and possibly throw an error if none is found.

Reply:
The validation is added at consumer level.

Reply:
Having no applications would be a valid business scenario if I am not wrong.